libslide/src/partial_evaluator/flatten.rs

```diff
@@ -0,0 +1,139 @@
+use crate::grammar::*;
+use crate::utils::{unflatten_binary_expr, UnflattenStrategy};
+
+use std::collections::{HashMap, VecDeque};
+use std::rc::Rc;
+
+pub fn flatten_expr(expr: &Rc<Expr>) -> Rc<Expr> {
+    match expr.as_ref() {
+        Expr::Const(_) | Expr::Var(_) => Rc::clone(expr),
+        Expr::Parend(inner) | Expr::Bracketed(inner) => Rc::clone(inner),
+        Expr::BinaryExpr(BinaryExpr { op, lhs, rhs })
+            if op == &BinaryOperator::Plus || op == &BinaryOperator::Minus =>
+        {
+            flatten_add_or_sub(lhs, rhs, op == &BinaryOperator::Minus)
+        }
+        // TODO: handle everything else
+        _ => Rc::clone(expr),
+    }
+}
+
+pub fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) -> Rc<Expr> {
+    let lhs = flatten_expr(o_lhs);
+    let rhs = flatten_expr(o_rhs);
+
+    let mut coeff = 0f64;
+    let mut terms = HashMap::<&Rc<Expr>, f64>::new();
+    let mut args = VecDeque::with_capacity(2);
+    let base_args = [lhs, rhs];
+    args.extend(base_args.iter());
+    let mut args_before_neg = 1;
```

Comment:
this deque could probably be created in one line;
I think rc should be an empty pointer, but Idk if default is implemented off my head
let mut args: VecDeque = vec![Rc<Expr>::default(), Rc<Expr>::default(), lhs, rhs]

Reply:
`with_capacity` reserves memory but doesn't populate anything. There is no macro for creating vecdeques in the standard library, but if you want you can roll your own: https://github.com/bluss/maplit/pull/31/files. That can be in a subsequent PR though.

Reply:
No, but you can just make a vector that easily casts into a deque. That’s how they do it on Rust docs

Reply:
Can you point me to documentation that does this? I'm not aware of anyway to do this

Reply:
https://doc.rust-lang.org/std/collections/struct.VecDeque.html
Reserve function shows them making one from a vec

Reply:
hmm..  i personally would rather keep it as is because I don't see a need to create and collect an iterator again. What do you think?

Reply:
That’s fine